src/main/java/javaslang/collection/List.java

```diff
@@ -426,7 +426,15 @@ default int lastIndexOf(T element) {
     @Override
     default Tuple2<List<T>, List<T>> partition(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        return Tuple.of(filter(predicate), filter(predicate.negate()));
+        List<T> left = nil(), right = nil();
+        for (T t : this) {
+            if (predicate.test(t)) {
+                left = left.prepend(t);
+            } else {
+                right = right.prepend(t);
+            }
+        }
+        return Tuple.of(left.reverse(), right.reverse());
```

Comment:
Note: List.ofAll(Interable) could be improved in a future version of Javaslang in the same way as for ArrayList by reverse-iterating all IndexedSeq instances in O(1). See https://github.com/javaslang/javaslang/issues/220#issuecomment-111876895